third_party/nvidia/lib/Dialect/NVWS/Transforms/LowerWarpGroup.cpp

```diff
@@ -0,0 +1,251 @@
+      });
+    }
+    return captures;
+  }
+
+  void populateBlock(PatternRewriter &rewriter, mlir::Block *input_block,
+                     mlir::Block *output_block, SmallVector<Value> &inputs,
+                     SmallVector<Value> &constants) const {
+
+    DenseMap<Value, Value> valueMap;
+    rewriter.setInsertionPointToEnd(output_block);
+    for (auto &value : inputs) {
+      auto new_value =
+          output_block->addArgument(value.getType(), value.getLoc());
+      valueMap[value] = new_value;
+    }
+    for (auto &value : constants) {
+      auto cOp = cast<arith::ConstantOp>(*value.getDefiningOp());
+      auto newConst =
+          rewriter.create<arith::ConstantOp>(value.getLoc(), cOp.getValue());
+      valueMap[value] = newConst->getResult(0);
+    }
+    auto retOp = rewriter.create<triton::gpu::WarpReturnOp>(
+        input_block->getParentOp()->getLoc(), ArrayRef<Type>(),
+        ArrayRef<Value>());
+
+    for (auto &op : llvm::make_early_inc_range(
+             llvm::make_range(input_block->getOperations().begin(),
+                              input_block->getTerminator()->getIterator()))) {
+      op.moveBefore(retOp);
+    }
+
+    auto traverseParentBlocks = [&](Operation *op) -> bool {
+      Block *currentBlock = op->getBlock();
+      while (currentBlock) {
+        // Check the current block
+        if (currentBlock == output_block)
+          return true;
+        // Get parent operation of this block
+        Operation *parentOp = currentBlock->getParentOp();
+        if (!parentOp)
+          break;
+        // Move up to parent block
+        currentBlock = parentOp->getBlock();
+      }
+      return false;
+    };
+
+    for (auto pair : valueMap) {
+      auto oldValue = pair.first;
+      auto newValue = pair.second;
+      rewriter.replaceUsesWithIf(oldValue, newValue, [&](OpOperand &operand) {
+        return traverseParentBlocks(operand.getOwner());
+      });
+    }
+  }
+
+  LogicalResult createTtgWSOp(Location loc, PatternRewriter &rewriter,
+                              Block *defaultPartition,
+                              ArrayRef<Block *> partitions,
+                              ArrayRef<int> numWarps,
+                              ArrayRef<int> warpGroupStartIds) const {
+    if (partitions.size() != numWarps.size())
+      return failure("mismatched number of warp groups and number of warps per "
+                     "warp group");
+    if (partitions.size() != warpGroupStartIds.size())
+      return failure(
+          "mismatched number of warp groups and number of warp start ids");
+    DenseSet<Value> uniqueInputs;
+    DenseSet<Value> uniqueConstants;
+    for (auto *block : partitions) {
+      for (Value &input : iterateInputs(block)) {
+        if (!isa<BlockArgument>(input) &&
+            isa<arith::ConstantOp>(input.getDefiningOp())) {
+          uniqueConstants.insert(input);
+        } else {
+          uniqueInputs.insert(input);
+        }
+      }
+    }
```

Comment:
I think you can replace this with `mlir::getUsedValuesDefinedAbove` which returns the implicit captures of a region. You can then sort them into block arguments and constants.

Reply:
Hmm, okay, will need to pass regions instead of blocks, but I can do that.